src/metabase/driver/h2.clj

```diff
@@ -127,22 +133,51 @@
       (when (instance? SessionLocal session)
         (Parser. session)))))
 
-(defn- contains-ddl?
+(mu/defn ^:private classify-query :- [:map
+                                      [:command-types [:vector pos-int?]]
+                                      [:remaining-sql [:maybe :string]]]
+  "Takes an h2 db id, and a query, returns the command-classes+types (see: ), and any remaining sql.
+
+  - Each `command-type` corresponds to a value in org.h2.command.CommandInterface, and match the commands in order.
+  - `remaining-sql` is a nillable sql string that is unable to be classified without running preceding queries first.
+    Usually if `remaining-sql` exists we will deny the query."
   [database query]
   (when-let [h2-parser (make-h2-parser database)]
     (try
-      (let [command      (.prepareCommand h2-parser query)
-            command-type (.getCommandType command)]
-        ;; TODO: do we need to handle CommandList?
-        ;; Command types are organized with all DDL commands listed first
-        ;; see https://github.com/h2database/h2database/blob/master/h2/src/main/org/h2/command/CommandInterface.java
-        (< command-type CommandInterface/ALTER_SEQUENCE))
-      ;; if the query is invalid, then it isn't DDL
-      (catch Throwable _ false))))
-
-(defn- check-disallow-ddl-commands [{:keys [database] {:keys [query]} :native}]
-  (when (and query (contains-ddl? database query))
-    (throw (IllegalArgumentException. "DDL commands are not allowed to be used with h2."))))
+      (let [command-list      (.prepareCommand h2-parser query)
+            command-list-type (.getCommandType ^org.h2.command.CommandList command-list)
```

Comment:
this typehint is wrong. `prepareCommand` returns a `org.h2.command.Command`. It can be either a `CommandContainer` or a `CommandList`. annotating with the list here is wrong in cases of a single statement in the sql.

But it's also not needed. If you remove it, the parser has type info, and Clojure infers the type of `command-list` from `prepareCommand`.

Reply:
It should probably be renamed from `command-list` to `command` too.

Reply:
Done